### PR TITLE
Added save as functionality

### PR DESCRIPTION
**Pull request by jkwok678**

### CODE REVIEW

File: canvas.cpp [hunk 1]
```diff
@@ -976,9 +976,14 @@ void Canvas::setGrid(bool newGrid)
...
-	map->checkAllTracksConnected();
+	return map->checkAllTracksConnected();
+}
+
+int Canvas::getTrackTotal()
+{
+	return map->getTotalTrack();
 }
 
 void Canvas::clickCreateAddElement(ElementChosen elementToAdd, int overallX, int overallY)
```
Comment (zarethrex) [inline canvas.cpp:989]: I am not a C++ maestro but would it not be better to pass by reference here, i.e. `ElementChosen&`? Depends how you are using the object. Also if the object is not being altered consider `const ElementChosen&`.

File: canvas.h [hunk 2]
```diff
@@ -235,7 +235,13 @@ class Canvas : public QWidget
...
 	 * @brief Checks all the tracks in the map class to see if they are all connected with no unconnected ends.
 	 */
-	void checkAllTracksInMapConnected();
+	bool checkAllTracksInMapConnected();
+
+	/**
+	 * @brief Get total number of tracks.
+	 * @return tracks Number of tracks as an int.
+	 */
+	int getTrackTotal();
```
Comment (zarethrex) [inline canvas.h]: Recommend `int getTrackTotal() const;`

Comment (zarethrex) [reply canvas.h]: (I see you did this in other places)

File: element.cpp [hunk 3]
```diff
@@ -203,6 +213,17 @@ QString NamedLocation::toQString()
 	return namedLocationQString;
 }
 
+QString NamedLocation::toQStringForSave()
+{
+	QString saveNamedLocationQString = locationToQString();
+	if(text != nullptr)
```
Comment (zarethrex) [inline element.cpp]: Should this not be just `if(!text)`?

File: element.cpp [hunk 4]
```diff
@@ -229,6 +250,17 @@ QString Concourse::toQString()
 	return concourseQString;
 }
 
+QString Concourse::toQStringForSave()
+{
+	QString saveConcourseQString = locationToQString();
+	if(text != nullptr)
```
Comment (zarethrex) [inline element.cpp]: Same as above

File: filesaver.cpp [hunk 5]
```diff
@@ -1,34 +1,85 @@
 #include "filesaver.h"
 
-Filesaver::Filesaver(QString newPath)
+//private
+
+std::vector<QString> Filesaver::prepareFileContentToWrite(Map *map)
```
Comment (zarethrex) [inline filesaver.cpp:5]: As `map` is not altered, maybe `const Map* map`?

File: filesaver.h [hunk 6]
```diff
@@ -9,11 +9,36 @@ class Filesaver
...
+	/**
+	 * @brief Filesaver default constructor.
+	 */
+	Filesaver();
+
+	/**
+	 * @brief Set a new file path for the filesaver.
+	 * @param newPath QString path.
+	 */
+	void setNewFilePath(QString newPath);
```
Comment (zarethrex) [inline filesaver.h:32]: Again I cannot claim to be an expert but I would use `const QString&`

File: filesaver.h [hunk 7]
```diff
@@ -9,11 +9,36 @@ class Filesaver
...
+
+	/**
+	 * @brief Save the railway to a file.
+	 * @param map The current map on screen.
+	 * @return true if it saved, false if it failed to save.
+	 */
+	bool saveRailway(Map *map);
```
Comment (zarethrex) [inline filesaver.h:39]: Is `map` altered here? Else `const Map*`

File: track.h [hunk 8]
```diff
@@ -704,16 +754,10 @@ class LinkedTrack : public Track
...
-	bool getLinked() const;
-
-	/**
-	 * @brief Set whether the track is linked.
-	 * @param newLinked A bool that says if the track is linked.
+	 * @brief Checks if the LinkedTrack is linked to another LinkedTrack.
+	 * @return true if it's linked otherwise false.
 	 */
-	void setLinked(bool newLinked);
+	bool isLinked();
```
Comment (zarethrex) [inline track.h]: `bool isLinked() const;` matching others

File: window.cpp [hunk 9]
```diff
@@ -143,14 +144,31 @@ void Window::loadRailway()
...
-	QString fileName = QFileDialog::getSaveFileName(this, tr("Save File"),
-							   "/home/jana/untitled.png",
-							   tr("Railway2 File (*.rly2)"));
+	QString fileName = "";
+	if (drawingArea->getTrackTotal() != 0)
+	{
+		if (drawingArea->checkAllTracksInMapConnected())
+		{
+			fileName = QFileDialog::getSaveFileName(this, tr("Save File"),
+								   "./",
```
Comment (zarethrex) [inline window.cpp]: Suggest not using hard coded paths, for current directory: `QDir::currentPath()`

File: window.cpp [hunk 10]
```diff
@@ -143,14 +144,31 @@ void Window::loadRailway()
...
+								   tr("Railway2 File (*.rly2);; Development2 File (*.dev2)"));
+		}
+		else
+		{
+			fileName = QFileDialog::getSaveFileName(this, tr("Save File"),
+								   "./",
```
Comment (zarethrex) [inline window.cpp]: See above

**Review by zarethrex — CHANGES_REQUESTED** [review]

File: canvas.h (hunk 2 above)
Comment (jkwok678) [reply canvas.h]: Yep probs laziness on my part, will fix.

File: element.cpp (hunk 3 above)
Comment (jkwok678) [reply element.cpp]: Can't be changed, change would crash the program

**Review by zarethrex — APPROVED** [review]

**Review by zarethrex — APPROVED** [review]

**Pull request merged**